In [1]:
import pandas as pd
import numpy as np
import requests
from bs4 import BeautifulSoup
import re
import json # library to handle JSON files

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library
from uszipcode import SearchEngine
import googlemaps
pd.set_option('display.max_rows', 500)

### Get list of 50 top college towns

In [2]:
url = "http://www.collegeranker.com/features/best-college-towns/"

In [3]:
r = requests.get(url)

In [4]:
soup = BeautifulSoup(r.text, "html.parser")

In [5]:
post_table = soup.find("div",class_="awr-i")

In [6]:
town_list = []
for row in post_table.find_all('h2'):
    town = row.text.split('.', 1)[1].lstrip()
    town_list.append(town)

In [7]:
# Dictionary to conver State Name to two letter code
us_state_abbrev = {
    'Alabama': 'AL',
    'Alaska': 'AK',
    'Arizona': 'AZ',
    'Arkansas': 'AR',
    'California': 'CA',
    'Colorado': 'CO',
    'Connecticut': 'CT',
    'Delaware': 'DE',
    'Florida': 'FL',
    'Georgia': 'GA',
    'Hawaii': 'HI',
    'Idaho': 'ID',
    'Illinois': 'IL',
    'Indiana': 'IN',
    'Iowa': 'IA',
    'Kansas': 'KS',
    'Kentucky': 'KY',
    'Louisiana': 'LA',
    'Maine': 'ME',
    'Maryland': 'MD',
    'Massachusetts': 'MA',
    'Michigan': 'MI',
    'Minnesota': 'MN',
    'Mississippi': 'MS',
    'Missouri': 'MO',
    'Montana': 'MT',
    'Nebraska': 'NE',
    'Nevada': 'NV',
    'New Hampshire': 'NH',
    'New Jersey': 'NJ',
    'New Mexico': 'NM',
    'New York': 'NY',
    'North Carolina': 'NC',
    'North Dakota': 'ND',
    'Ohio': 'OH',
    'Oklahoma': 'OK',
    'Oregon': 'OR',
    'Pennsylvania': 'PA',
    'Rhode Island': 'RI',
    'South Carolina': 'SC',
    'South Dakota': 'SD',
    'Tennessee': 'TN',
    'Texas': 'TX',
    'Utah': 'UT',
    'Vermont': 'VT',
    'Virginia': 'VA',
    'Washington': 'WA',
    'West Virginia': 'WV',
    'Wisconsin': 'WI',
    'Wyoming': 'WY',
}

In [8]:
town_df = pd.DataFrame([(x.split(',')[0], x.split(',')[1].lstrip()) for x in town_list])
town_df.columns = ['City', 'State']

In [9]:
# Convert States to Two-Letter Code
town_df['State'] = [us_state_abbrev[x] if x in list(us_state_abbrev.keys()) else x for x in town_df['State']]

In [10]:
town_df['Town'] = town_df['City'] + ', ' + town_df['State'].map(str)

### Use geocoder to get latitiude and longitude

In [11]:
def get_coordinates(df):
    lat_list = []
    lng_list = []
    API_KEY = 'AIzaSyDfaz-Hluid0RfZy6NDa3z9l26Hcuws93M'
    gmaps = googlemaps.Client(key=API_KEY)
    for x in df['Town']:
        geocode_result = gmaps.geocode(x)
        lat_list.append(geocode_result[0]['geometry']['location']['lat'])
        lng_list.append(geocode_result[0]['geometry']['location']['lng'])
    df['lat'] = lat_list
    df['lng'] = lng_list
    return df

In [12]:
get_coordinates(town_df)

City State                 Town        lat         lng
0              Troy    AL             Troy, AL  31.808768  -85.969951
1         Wellesley    MA        Wellesley, MA  42.296797  -71.292388
2          Columbus    MS         Columbus, MS  33.495674  -88.427263
3          Berkeley    CA         Berkeley, CA  37.871593 -122.272747
4            Lawton    OK           Lawton, OK  34.603567  -98.395929
5        Vermillion    SD       Vermillion, SD  42.779442  -96.929210
6         Brunswick    GA        Brunswick, GA  31.149953  -81.491489
7      Williamsburg    VA     Williamsburg, VA  37.270702  -76.707457
8           Boulder    CO          Boulder, CO  40.014986 -105.270546
9            Monroe    LA           Monroe, LA  32.509311  -92.119301
10  Charlottesville    VA  Charlottesville, VA  38.029306  -78.476678
11      Gainesville    FL      Gainesville, FL  29.651634  -82.324826
12       Burlington    VT       Burlington, VT  44.475882  -73.212072
13       Long Beach    CA       Long Beach, CA  33.770050 -118.193740
14        Flagstaff    AZ        Flagstaff, AZ  35.198284 -111.651302
15           Ithaca    NY           Ithaca, NY  42.443961  -76.501881
16           Durham    NC           Durham, NC  35.994033  -78.898619
17           Auburn    AL           Auburn, AL  32.609857  -85.480782
18         Savannah    GA         Savannah, GA  32.080899  -81.091203
19      Panama City    FL      Panama City, FL  30.158813  -85.660206
20     East Lansing    MI     East Lansing, MI  42.736979  -84.483865
21           Eugene    OR           Eugene, OR  44.052069 -123.086754
22    State College    PA    State College, PA  40.793395  -77.860001
23     Fayetteville    AR     Fayetteville, AR  36.082156  -94.171854
24        Manhattan    KS        Manhattan, KS  39.183608  -96.571669
25            Logan    UT            Logan, UT  41.736980 -111.833836
26       Morgantown    WV       Morgantown, WV  39.629526  -79.955897
27        Asheville    NC        Asheville, NC  35.595058  -82.551487
28           Duluth    MN           Duluth, MN  46.786672  -92.100485
29       Santa Cruz    CA       Santa Cruz, CA  36.974117 -122.030796
30        Ann Arbor    MI        Ann Arbor, MI  42.280826  -83.743038
31         Syracuse    NY         Syracuse, NY  43.048122  -76.147424
32            Fargo    ND            Fargo, ND  46.877186  -96.789803
33           Athens    GA           Athens, GA  33.951935  -83.357567
34          Madison    WI          Madison, WI  43.073052  -89.401230
35            Provo    UT            Provo, UT  40.233844 -111.658534
36          Bozeman    MT          Bozeman, MT  45.676998 -111.042934
37           Oxford    MS           Oxford, MS  34.366495  -89.519248
38          Lincoln    NE          Lincoln, NE  40.813616  -96.702596
39        Corvallis    OR        Corvallis, OR  44.564566 -123.262044
40         Beaufort    SC         Beaufort, SC  32.431581  -80.669829
41           Malibu    CA           Malibu, CA  34.025922 -118.779757
42            Tempe    AZ            Tempe, AZ  33.425510 -111.940005
43       South Bend    IN       South Bend, IN  41.676355  -86.251990
44      Bloomington    IN      Bloomington, IN  39.165325  -86.526386
45             Ames    IA             Ames, IA  42.030781  -93.631913
46           Athens    OH           Athens, OH  39.329240  -82.101255
47     Fort Collins    CO     Fort Collins, CO  40.585260 -105.084423
48         Lawrence    KS         Lawrence, KS  38.971669  -95.235250
49    St. Augustine    FL    St. Augustine, FL  29.901244  -81.312434

### Get Foursquare data on each town

In [13]:
# Credentials for URL Scrape
CLIENT_ID = 'LI4K1OOUAFPJJSTNNZ23CZ05ROFAGZ5LLOOM1WNWWE0PN0P0' # your Foursquare ID
CLIENT_SECRET = 'P1N0RHNW1IMWEVOT205NDGHN1MIGCGAKN52HCH3JVTZ4AVRF' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 200 # limit of number of venues returned by Foursquare API

In [14]:
def getNearbyVenues(towns):
    count = 1
    venues_list=[]
    for town in towns:
        print(town, count)
        count += 1
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&near={}&limit={}'.format(
                    CLIENT_ID, 
                    CLIENT_SECRET, 
                    VERSION, 
                    town,
                    LIMIT)

        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(x['venue']['name'],
                            x['venue']['location']['lat'], 
                            x['venue']['location']['lng'],
                            town,
                            x['venue']['categories'][0]['name']) for x in results])

        nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
        nearby_venues.columns = ['Venue',
                                'Venue Latitude', 
                                'Venue Longitude',
                                'City',
                                'Venue Category']
        
    return(nearby_venues)

In [15]:
all_venues = getNearbyVenues(town_df['Town'])

Troy, AL 1
Wellesley, MA 2
Columbus, MS 3
Berkeley, CA 4
Lawton, OK 5
Vermillion, SD 6
Brunswick, GA 7
Williamsburg, VA 8
Boulder, CO 9
Monroe, LA 10
Charlottesville, VA 11
Gainesville, FL 12
Burlington, VT 13
Long Beach, CA 14
Flagstaff, AZ 15
Ithaca, NY 16
Durham, NC 17
Auburn, AL 18
Savannah, GA 19
Panama City, FL 20
East Lansing, MI 21
Eugene, OR 22
State College, PA 23
Fayetteville, AR 24
Manhattan, KS 25
Logan, UT 26
Morgantown, WV 27
Asheville, NC 28
Duluth, MN 29
Santa Cruz, CA 30
Ann Arbor, MI 31
Syracuse, NY 32
Fargo, ND 33
Athens, GA 34
Madison, WI 35
Provo, UT 36
Bozeman, MT 37
Oxford, MS 38
Lincoln, NE 39
Corvallis, OR 40
Beaufort, SC 41
Malibu, CA 42
Tempe, AZ 43
South Bend, IN 44
Bloomington, IN 45
Ames, IA 46
Athens, OH 47
Fort Collins, CO 48
Lawrence, KS 49
St. Augustine, FL 50


### Map of all locations

In [16]:
map_df = all_venues.groupby('City').mean().reset_index()
map_df.columns = ['City', 'Latitude', 'Longitude']

In [17]:
map_df.head()

City   Latitude  Longitude
0       Ames, IA  42.023742 -93.632158
1  Ann Arbor, MI  42.273732 -83.745716
2  Asheville, NC  35.574073 -82.553108
3     Athens, GA  33.946499 -83.390763
4     Athens, OH  39.331427 -82.088832

In [18]:
# create map of US using latitude and longitude values
map_us = folium.Map(location=[48, -102], zoom_start=4)

# add markers to map
for lat, lng, city in zip(town_df['lat'], town_df['lng'], town_df['City']):
    label = '{}'.format(city)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7).add_to(map_us)  
    
map_us

### Analyze each neighborhood

In [19]:
# one hot encoding
venues_onehot = pd.get_dummies(all_venues[['Venue Category']], prefix="", prefix_sep="")

# add city and state column back to dataframe
venues_onehot['City'] = all_venues['City']

# move city column to the first column
fixed_columns = [venues_onehot.columns[-1]] + list(venues_onehot.columns[:-1])
venues_onehot = venues_onehot[fixed_columns]

venues_onehot.head()

City  Accessories Store  African Restaurant  Airport  \
0  Troy, AL                  0                   0        0   
1  Troy, AL                  0                   0        0   
2  Troy, AL                  0                   0        0   
3  Troy, AL                  0                   0        0   
4  Troy, AL                  0                   0        0   

   American Restaurant  Amphitheater  Animal Shelter  Antique Shop  Aquarium  \
0                    0             0               0             0         0   
1                    0             0               0             0         0   
2                    0             0               0             0         0   
3                    0             0               0             0         0   
4                    0             0               0             0         0   

   Arcade ...   Water Park  Waterfall  Whisky Bar  Wine Bar  Wine Shop  \
0       0 ...            0          0           0         0          0   
1       0 ...            0          0           0         0          0   
2       0 ...            0          0           0         0          0   
3       0 ...            0          0           0         0          0   
4       0 ...            0          0           0         0          0   

   Winery  Wings Joint  Women's Store  Yoga Studio  Zoo  
0       0            0              0            0    0  
1       0            0              0            0    0  
2       0            0              0            0    0  
3       0            0              0            0    0  
4       0            0              0            0    0  

[5 rows x 307 columns]

In [20]:
venues_grouped = venues_onehot.groupby('City').mean().reset_index()
venues_grouped

City  Accessories Store  African Restaurant  Airport  \
0              Ames, IA           0.000000                0.00     0.00   
1         Ann Arbor, MI           0.000000                0.00     0.00   
2         Asheville, NC           0.000000                0.00     0.00   
3            Athens, GA           0.000000                0.00     0.00   
4            Athens, OH           0.000000                0.00     0.00   
5            Auburn, AL           0.000000                0.00     0.00   
6          Beaufort, SC           0.000000                0.00     0.00   
7          Berkeley, CA           0.000000                0.00     0.00   
8       Bloomington, IN           0.000000                0.00     0.00   
9           Boulder, CO           0.000000                0.00     0.00   
10          Bozeman, MT           0.000000                0.00     0.00   
11        Brunswick, GA           0.000000                0.00     0.00   
12       Burlington, VT           0.010000                0.00     0.00   
13  Charlottesville, VA           0.000000                0.01     0.00   
14         Columbus, MS           0.012346                0.00     0.00   
15        Corvallis, OR           0.000000                0.00     0.00   
16           Duluth, MN           0.000000                0.00     0.00   
17           Durham, NC           0.000000                0.00     0.00   
18     East Lansing, MI           0.000000                0.00     0.00   
19           Eugene, OR           0.000000                0.00     0.00   
20            Fargo, ND           0.000000                0.00     0.00   
21     Fayetteville, AR           0.000000                0.00     0.00   
22        Flagstaff, AZ           0.000000                0.00     0.00   
23     Fort Collins, CO           0.000000                0.00     0.00   
24      Gainesville, FL           0.000000                0.00     0.00   
25           Ithaca, NY           0.000000                0.00     0.00   
26         Lawrence, KS           0.000000                0.00     0.00   
27           Lawton, OK           0.011494                0.00     0.00   
28          Lincoln, NE           0.000000                0.00     0.00   
29            Logan, UT           0.000000                0.00     0.00   
30       Long Beach, CA           0.000000                0.00     0.00   
31          Madison, WI           0.000000                0.00     0.00   
32           Malibu, CA           0.000000                0.00     0.00   
33        Manhattan, KS           0.000000                0.00     0.00   
34           Monroe, LA           0.010753                0.00     0.00   
35       Morgantown, WV           0.000000                0.00     0.00   
36           Oxford, MS           0.000000                0.00     0.00   
37      Panama City, FL           0.000000                0.00     0.00   
38            Provo, UT           0.000000                0.00     0.00   
39       Santa Cruz, CA           0.000000                0.00     0.00   
40         Savannah, GA           0.010000                0.01     0.00   
41       South Bend, IN           0.000000                0.00     0.00   
42    St. Augustine, FL           0.010000                0.00     0.01   
43    State College, PA           0.000000                0.00     0.00   
44         Syracuse, NY           0.000000                0.00     0.00   
45            Tempe, AZ           0.000000                0.00     0.00   
46             Troy, AL           0.000000                0.00     0.00   
47       Vermillion, SD           0.000000                0.00     0.00   
48        Wellesley, MA           0.000000                0.00     0.00   
49     Williamsburg, VA           0.010000                0.00     0.00   

    American Restaurant  Amphitheater  Animal Shelter  Antique Shop  Aquarium  \
0              0.021505          0.00            0.00          0.00      0.00   
1              0.020000        

In [21]:
len(venues_grouped)

50

In [22]:
num_top_venues = 10

for town in venues_grouped['City']:
    print("----"+town+"----")
    temp = venues_grouped[venues_grouped['City'] == town].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Ames, IA----
                  venue  freq
0           Pizza Place  0.09
1                   Bar  0.08
2         Grocery Store  0.05
3                  Café  0.05
4           Coffee Shop  0.04
5            Restaurant  0.03
6        Sandwich Place  0.03
7  Fast Food Restaurant  0.02
8                  Park  0.02
9  Gym / Fitness Center  0.02


----Ann Arbor, MI----
            venue  freq
0   Grocery Store  0.06
1     Coffee Shop  0.05
2     Pizza Place  0.04
3  Ice Cream Shop  0.04
4             Bar  0.03
5            Park  0.03
6         Brewery  0.03
7     Music Venue  0.03
8      Steakhouse  0.02
9  Sandwich Place  0.02


----Asheville, NC----
                 venue  freq
0              Brewery  0.13
1  American Restaurant  0.05
2                Hotel  0.04
3        Grocery Store  0.04
4       Breakfast Spot  0.04
5                  Bar  0.04
6          Coffee Shop  0.03
7               Garden  0.03
8            BBQ Joint  0.03
9                 Park  0.02


----Athens, GA----
 

                 venue  freq
0        Grocery Store  0.10
1   Mexican Restaurant  0.05
2  American Restaurant  0.05
3          Pizza Place  0.05
4          Coffee Shop  0.04
5                  Bar  0.04
6   Italian Restaurant  0.04
7            BBQ Joint  0.04
8       Breakfast Spot  0.03
9       Sandwich Place  0.03


----Ithaca, NY----
                     venue  freq
0              Coffee Shop  0.05
1                     Park  0.04
2           Ice Cream Shop  0.04
3          Thai Restaurant  0.03
4           Sandwich Place  0.03
5                   Bakery  0.03
6               Bagel Shop  0.03
7  New American Restaurant  0.03
8      American Restaurant  0.03
9       Italian Restaurant  0.03


----Lawrence, KS----
                venue  freq
0         Coffee Shop  0.09
1        Burger Joint  0.05
2                 Bar  0.05
3         Pizza Place  0.04
4  Mexican Restaurant  0.04
5      Breakfast Spot  0.03
6             Brewery  0.03
7      Ice Cream Shop  0.03
8      Sandwich Place 

In [23]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [24]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['City']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
town_venues_sorted = pd.DataFrame(columns=columns)
town_venues_sorted['City'] = venues_grouped['City']

for ind in np.arange(venues_grouped.shape[0]):
    town_venues_sorted.iloc[ind, 1:] = return_most_common_venues(venues_grouped.iloc[ind, :], num_top_venues)

town_venues_sorted

City 1st Most Common Venue         2nd Most Common Venue  \
0              Ames, IA           Pizza Place                           Bar   
1         Ann Arbor, MI         Grocery Store                   Coffee Shop   
2         Asheville, NC               Brewery           American Restaurant   
3            Athens, GA    Mexican Restaurant                           Bar   
4            Athens, OH           Pizza Place                   Coffee Shop   
5            Auburn, AL   American Restaurant            Mexican Restaurant   
6          Beaufort, SC   American Restaurant                         Hotel   
7          Berkeley, CA                  Park                   Coffee Shop   
8       Bloomington, IN               Brewery           American Restaurant   
9           Boulder, CO                 Trail                       Brewery   
10          Bozeman, MT           Coffee Shop                           Bar   
11        Brunswick, GA   American Restaurant                Sandwich Place   
12       Burlington, VT           Pizza Place                       Brewery   
13  Charlottesville, VA                 Hotel            Mexican Restaurant   
14         Columbus, MS        Discount Store          Fast Food Restaurant   
15        Corvallis, OR           Coffee Shop                          Park   
16           Duluth, MN           Coffee Shop           American Restaurant   
17           Durham, NC            Beer Store                  Burger Joint   
18     East Lansing, MI      Sushi Restaurant                   Pizza Place   
19           Eugene, OR               Brewery                   Pizza Place   
20            Fargo, ND           Coffee Shop                   Pizza Place   
21     Fayetteville, AR          Burger Joint                   Coffee Shop   
22        Flagstaff, AZ           Coffee Shop                       Brewery   
23     Fort Collins, CO               Brewery                   Coffee Shop   
24      Gainesville, FL         Grocery Store           American Restaurant   
25           Ithaca, NY           Coffee Shop                Ice Cream Shop   
26         Lawrence, KS           Coffee Shop                  Burger Joint   
27           Lawton, OK  Fast Food Restaurant            Mexican Restaurant   
28          Lincoln, NE           Coffee Shop                 Grocery Store   
29            Logan, UT           Pizza Place            Mexican Restaurant   
30       Long Beach, CA                 Beach           American Restaurant   
31          Madison, WI           Pizza Place                  Liquor Store   
32           Malibu, CA                 Beach                     Surf Spot   
33        Manhattan, KS           Coffee Shop           American Restaurant   
34           Monroe, LA    Seafood Restaurant          Fast Food Restaurant   
35       Morgantown, WV           Coffee Shop                    Restaurant   
36           Oxford, MS           Coffee Shop                Sandwich Place   
37      Panama City, FL    Seafood Restaurant          Fast Food Restaurant   
38            Provo, UT        Sandwich Place            Mexican Restaurant   
39       Santa Cruz, CA                 Beach                   Coffee Shop   
40         Savannah, GA                 Plaza                        Museum   
41       South Bend, IN    Chinese Restaurant            Italian Restaurant   
42    St. Augustine, FL           Pizza Place           American Restaurant   
43    State College, PA   American Restaurant                Breakfast Spot   
44         Syracuse, NY           Pizza Place                        Bakery   
45            Tempe, AZ           Coffee Shop           American Restaurant   
46             Troy, AL  Fast Food Restaurant           American Restaurant   
47       Vermillion, SD           Pizza Place          Fast Food Restaurant   
48        Wellesley, MA           Pizza Place            Italian Restaurant   
49     Williamsburg, VA            Theme Park  Theme Park Ride / Attr

### Cluster neighborhoods

In [25]:
# set number of clusters
kclusters = 5

venues_grouped_clustering = venues_grouped.drop('City', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(venues_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([4, 2, 0, 4, 4, 4, 1, 2, 4, 2], dtype=int32)

### Evaluate Clusters

In [45]:
# Fit several clustering models that take n_clusters
from sklearn import cluster
from sklearn.preprocessing import StandardScaler

X = StandardScaler().fit_transform(venues_grouped_clustering)

models = [cluster.SpectralClustering(n_clusters=5), 
          cluster.KMeans(n_clusters=5), 
          cluster.MiniBatchKMeans(n_clusters=5), 
          cluster.AgglomerativeClustering(n_clusters=5)]

for model in models:
    model.fit(X)
    print(model.__class__.__name__)
    print("\tFirst 10 labels", model.labels_[:10])
    print("\t", len(model.labels_))

SpectralClustering
	First 10 labels [0 0 0 0 0 4 0 0 0 0]
	 50
KMeans
	First 10 labels [1 3 3 1 1 1 1 1 1 3]
	 50
MiniBatchKMeans
	First 10 labels [3 3 3 3 3 3 3 3 3 3]
	 50
AgglomerativeClustering
	First 10 labels [0 0 0 0 0 0 0 0 0 0]
	 50


In [46]:
# ... then measure their pairwise similarity
from sklearn import metrics
from itertools import combinations

for clust1, clust2 in combinations(models, 2):
    print(clust1.__class__.__name__, "versus", clust2.__class__.__name__)
    print("\tRand score:", metrics.adjusted_rand_score(clust1.labels_, clust2.labels_))
    print("\tMutual info:", metrics.adjusted_mutual_info_score(clust1.labels_, clust2.labels_))

SpectralClustering versus KMeans
	Rand score: -0.0736862516967229
	Mutual info: -0.04492752084537756
SpectralClustering versus MiniBatchKMeans
	Rand score: -0.11041258516330876
	Mutual info: -0.06157292834619116
SpectralClustering versus AgglomerativeClustering
	Rand score: -0.10679660930917206
	Mutual info: -0.06668319136534008
KMeans versus MiniBatchKMeans
	Rand score: 0.03250599631607545
	Mutual info: 0.13717046666222324
KMeans versus AgglomerativeClustering
	Rand score: 0.017655255823736705
	Mutual info: 0.1438674688851498
MiniBatchKMeans versus AgglomerativeClustering
	Rand score: 0.31332365779680815
	Mutual info: 0.21853287747643182


### Visualize Clusters

In [27]:
venues_merged = town_df[['Town', 'lat', 'lng']]
venues_merged.columns = ['City', 'Latitude', 'Longitude']
# add clustering labels
venues_merged['Cluster Labels'] = kmeans.labels_

# merge venues_grouped with to add latitude/longitude for each neighborhood
venues_merged = venues_merged.join(town_venues_sorted.set_index('City'), on='City')

venues_merged.head() # check the last columns!

/usr/local/Cellar/ipython/6.5.0/libexec/vendor/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  after removing the cwd from sys.path.


City   Latitude   Longitude  Cluster Labels 1st Most Common Venue  \
0       Troy, AL  31.808768  -85.969951               4  Fast Food Restaurant   
1  Wellesley, MA  42.296797  -71.292388               2           Pizza Place   
2   Columbus, MS  33.495674  -88.427263               0        Discount Store   
3   Berkeley, CA  37.871593 -122.272747               4                  Park   
4     Lawton, OK  34.603567  -98.395929               4  Fast Food Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0   American Restaurant        Sandwich Place              Pharmacy   
1    Italian Restaurant            Donut Shop           Coffee Shop   
2  Fast Food Restaurant   American Restaurant        Sandwich Place   
3           Coffee Shop        Ice Cream Shop               Brewery   
4    Mexican Restaurant          Burger Joint   American Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0           Pizza Place    Mexican Restaurant                 Hotel   
1       Thai Restaurant   Japanese Restaurant         Grocery Store   
2           Coffee Shop    Mexican Restaurant            Steakhouse   
3   Japanese Restaurant                 Trail    Mexican Restaurant   
4        Ice Cream Shop             BBQ Joint           Pizza Place   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0             BBQ Joint        Discount Store          Grocery Store  
1    Chinese Restaurant              Pharmacy                 Bakery  
2                 Hotel             BBQ Joint    Sporting Goods Shop  
3               Theater       Thai Restaurant            Flower Shop  
4        Cosmetics Shop        Sandwich Place         Clothing Store

In [28]:
# create map
map_clusters = folium.Map(location=[48, -102], zoom_start=4)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(venues_merged['Latitude'], venues_merged['Longitude'], venues_merged['City'], venues_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examine Clusters

In [29]:
venues_merged.loc[venues_merged['Cluster Labels'] == 0]

City   Latitude   Longitude  Cluster Labels  \
2          Columbus, MS  33.495674  -88.427263               0   
10  Charlottesville, VA  38.029306  -78.476678               0   
15           Ithaca, NY  42.443961  -76.501881               0   
16           Durham, NC  35.994033  -78.898619               0   
20     East Lansing, MI  42.736979  -84.483865               0   
22    State College, PA  40.793395  -77.860001               0   
23     Fayetteville, AR  36.082156  -94.171854               0   
26       Morgantown, WV  39.629526  -79.955897               0   
33           Athens, GA  33.951935  -83.357567               0   
39        Corvallis, OR  44.564566 -123.262044               0   
45             Ames, IA  42.030781  -93.631913               0   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
2         Discount Store  Fast Food Restaurant   American Restaurant   
10                 Hotel    Mexican Restaurant   American Restaurant   
15           Coffee Shop        Ice Cream Shop                  Park   
16            Beer Store          Burger Joint        Sandwich Place   
20      Sushi Restaurant           Pizza Place           Coffee Shop   
22   American Restaurant        Breakfast Spot           Pizza Place   
23          Burger Joint           Coffee Shop                Bakery   
26           Coffee Shop            Restaurant  Fast Food Restaurant   
33    Mexican Restaurant                   Bar           Pizza Place   
39           Coffee Shop                  Park        Sandwich Place   
45           Pizza Place                   Bar         Grocery Store   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
2         Sandwich Place           Coffee Shop    Mexican Restaurant   
10           Pizza Place     French Restaurant            Bagel Shop   
15    Italian Restaurant       Thai Restaurant                Bakery   
16               Brewery             Gastropub           Coffee Shop   
20                   Bar        Sandwich Place   American Restaurant   
22                  Park           Gas Station                  Café   
23  Fast Food Restaurant               Brewery        Sandwich Place   
26   American Restaurant           Gas Station                   Bar   
33  Fast Food Restaurant           Coffee Shop         Grocery Store   
39           Pizza Place               Brewery         Grocery Store   
45                  Café           Coffee Shop            Restaurant   

   7th Most Common Venue 8th Most Common Venue          9th Most Common Venue  \
2             Steakhouse                 Hotel                      BBQ Joint   
10             BBQ Joint        Sandwich Place                        Brewery   
15            Bagel Shop        Sandwich Place        New American Restaurant   
16        Ice Cream Shop                 Hotel                   Concert Hall   
20            Restaurant        Ice Cream Shop                    Golf Course   
22                 Hotel        Sandwich Place             Salon / Barbershop   
23           Pizza Place             BBQ Joint                  Grocery Store   
26           Pizza Place    Italian Restaurant             Mexican Restaurant   
33   American Restaurant          Liquor Store                        Brewery   
39                Bakery    Mexican Restaurant  Vegetarian / Vegan Restaurant   
45        Sandwich Place                   Gym                      BBQ Joint   

     10th Most Common Venue  
2       Sporting Goods Shop  
10             Burger Joint  
15      American Restaurant  
16                   Bakery  
20                Juice Bar  
22                      Bar  
23      Fried Chicken Joint  
26                    Trail  
33  New American Restaurant  
39             Burger Joint  
45       Mexican Restaurant

In [30]:
venues_merged.loc[venues_merged['Cluster Labels'] == 1]

City   Latitude   Longitude  Cluster Labels  \
6   Brunswick, GA  31.149953  -81.491489               1   
14  Flagstaff, AZ  35.198284 -111.651302               1   
27  Asheville, NC  35.595058  -82.551487               1   
34    Madison, WI  43.073052  -89.401230               1   
37     Oxford, MS  34.366495  -89.519248               1   
46     Athens, OH  39.329240  -82.101255               1   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
6    American Restaurant        Sandwich Place         Grocery Store   
14           Coffee Shop               Brewery           Pizza Place   
27               Brewery   American Restaurant         Grocery Store   
34           Pizza Place          Liquor Store   American Restaurant   
37           Coffee Shop        Sandwich Place        Discount Store   
46           Pizza Place           Coffee Shop              Dive Bar   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
6     Mexican Restaurant             BBQ Joint           Pizza Place   
14   American Restaurant    Mexican Restaurant                 Hotel   
27        Breakfast Spot                   Bar                 Hotel   
34             Gastropub                  Park               Brewery   
37            Restaurant           Pizza Place                 Hotel   
46                 Hotel                   Bar    Mexican Restaurant   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
6       Department Store              Pharmacy        Clothing Store   
14            Restaurant          Dessert Shop   Sporting Goods Shop   
27           Coffee Shop             BBQ Joint                Garden   
34         Grocery Store                  Café           Coffee Shop   
37             Bookstore   American Restaurant                   Bar   
46    Chinese Restaurant        Sandwich Place              Beer Bar   

   10th Most Common Venue  
6             Coffee Shop  
14           Burger Joint  
27               Wine Bar  
34             Restaurant  
37                 Bakery  
46    American Restaurant

In [31]:
venues_merged.loc[venues_merged['Cluster Labels'] == 2]

City   Latitude   Longitude  Cluster Labels  \
1       Wellesley, MA  42.296797  -71.292388               2   
7    Williamsburg, VA  37.270702  -76.707457               2   
9          Monroe, LA  32.509311  -92.119301               2   
12     Burlington, VT  44.475882  -73.212072               2   
17         Auburn, AL  32.609857  -85.480782               2   
19    Panama City, FL  30.158813  -85.660206               2   
25          Logan, UT  41.736980 -111.833836               2   
30      Ann Arbor, MI  42.280826  -83.743038               2   
31       Syracuse, NY  43.048122  -76.147424               2   
40       Beaufort, SC  32.431581  -80.669829               2   
42          Tempe, AZ  33.425510 -111.940005               2   
49  St. Augustine, FL  29.901244  -81.312434               2   

   1st Most Common Venue         2nd Most Common Venue 3rd Most Common Venue  \
1            Pizza Place            Italian Restaurant            Donut Shop   
7             Theme Park  Theme Park Ride / Attraction    Seafood Restaurant   
9     Seafood Restaurant          Fast Food Restaurant    Mexican Restaurant   
12           Pizza Place                       Brewery                Bakery   
17   American Restaurant            Mexican Restaurant           Coffee Shop   
19    Seafood Restaurant          Fast Food Restaurant    Italian Restaurant   
25           Pizza Place            Mexican Restaurant        Sandwich Place   
30         Grocery Store                   Coffee Shop           Pizza Place   
31           Pizza Place                        Bakery    Italian Restaurant   
40   American Restaurant                         Hotel         Grocery Store   
42           Coffee Shop           American Restaurant        Sandwich Place   
49           Pizza Place           American Restaurant        Ice Cream Shop   

   4th Most Common Venue          5th Most Common Venue 6th Most Common Venue  \
1            Coffee Shop                Thai Restaurant   Japanese Restaurant   
7         History Museum                      Gift Shop           Coffee Shop   
9         Sandwich Place            American Restaurant      Department Store   
12           Coffee Shop  Vegetarian / Vegan Restaurant        Sandwich Place   
17             BBQ Joint                  Grocery Store        Sandwich Place   
19        Sandwich Place            American Restaurant         Grocery Store   
25  Fast Food Restaurant                    Coffee Shop   American Restaurant   
30        Ice Cream Shop                        Brewery                   Bar   
31                   Pub                          Diner           Coffee Shop   
40    Seafood Restaurant                 Discount Store  Fast Food Restaurant   
42        Breakfast Spot                    Pizza Place                   Bar   
49                  Café             Seafood Restaurant           Coffee Shop   

   7th Most Common Venue 8th Most Common Venue      9th Most Common Venue  \
1          Grocery Store    Chinese Restaurant                   Pharmacy   
7   Fast Food Restaurant         Historic Site                Golf Course   
9            Coffee Shop        Cosmetics Shop  Cajun / Creole Restaurant   
12                  Café                  Park                        Gym   
17              Pharmacy           Pizza Place               Burger Joint   
19        Breakfast Spot           Pizza Place                  BBQ Joint   
25                Bakery          Burger Joint             Ice Cream Shop   
30                  Park           Music Venue             Sandwich Place   
31        Ice Cream Shop        Clothing Store                        Bar   
40             BBQ Joint                   Bar             Sandwich Place   
42     Convenience Store                  Park              Grocery Store   
49         Grocery Store          Cocktail Bar       Caribbean Restaurant   

   10th Most Common Venue  
1                  Bakery  
7     American Restaurant  
9  

In [32]:
venues_merged.loc[venues_merged['Cluster Labels'] == 3]

City   Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
32  Fargo, ND  46.877186 -96.789803               3           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
32           Pizza Place          Burger Joint   American Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
32        Sandwich Place    Mexican Restaurant                Bakery   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
32               Brewery                   Gym       Asian Restaurant

In [33]:
venues_merged.loc[venues_merged['Cluster Labels'] == 4]

City   Latitude   Longitude  Cluster Labels  \
0           Troy, AL  31.808768  -85.969951               4   
3       Berkeley, CA  37.871593 -122.272747               4   
4         Lawton, OK  34.603567  -98.395929               4   
5     Vermillion, SD  42.779442  -96.929210               4   
8        Boulder, CO  40.014986 -105.270546               4   
11   Gainesville, FL  29.651634  -82.324826               4   
13    Long Beach, CA  33.770050 -118.193740               4   
18      Savannah, GA  32.080899  -81.091203               4   
21        Eugene, OR  44.052069 -123.086754               4   
24     Manhattan, KS  39.183608  -96.571669               4   
28        Duluth, MN  46.786672  -92.100485               4   
29    Santa Cruz, CA  36.974117 -122.030796               4   
35         Provo, UT  40.233844 -111.658534               4   
36       Bozeman, MT  45.676998 -111.042934               4   
38       Lincoln, NE  40.813616  -96.702596               4   
41        Malibu, CA  34.025922 -118.779757               4   
43    South Bend, IN  41.676355  -86.251990               4   
44   Bloomington, IN  39.165325  -86.526386               4   
47  Fort Collins, CO  40.585260 -105.084423               4   
48      Lawrence, KS  38.971669  -95.235250               4   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0   Fast Food Restaurant   American Restaurant        Sandwich Place   
3                   Park           Coffee Shop        Ice Cream Shop   
4   Fast Food Restaurant    Mexican Restaurant          Burger Joint   
5            Pizza Place  Fast Food Restaurant                   Bar   
8                  Trail               Brewery        Ice Cream Shop   
11         Grocery Store   American Restaurant           Pizza Place   
13                 Beach   American Restaurant         Grocery Store   
18                 Plaza                Museum                  Café   
21               Brewery           Pizza Place                  Park   
24           Coffee Shop   American Restaurant         Grocery Store   
28           Coffee Shop   American Restaurant                  Park   
29                 Beach           Coffee Shop         Grocery Store   
35        Sandwich Place    Mexican Restaurant                 Trail   
36           Coffee Shop                   Bar   American Restaurant   
38           Coffee Shop         Grocery Store   Fried Chicken Joint   
41                 Beach             Surf Spot                 Trail   
43    Chinese Restaurant    Italian Restaurant         Grocery Store   
44               Brewery   American Restaurant  Gym / Fitness Center   
47               Brewery           Coffee Shop           Pizza Place   
48           Coffee Shop          Burger Joint                   Bar   

              4th Most Common Venue 5th Most Common Venue  \
0                          Pharmacy           Pizza Place   
3                           Brewery   Japanese Restaurant   
4               American Restaurant        Ice Cream Shop   
5                             Hotel        Sandwich Place   
8                    Sandwich Place   American Restaurant   
11               Mexican Restaurant    Italian Restaurant   
13                   Ice Cream Shop    Seafood Restaurant   
18  Southern / Soul Food Restaurant           Coffee Shop   
21                      Coffee Shop        Breakfast Spot   
24                   Breakfast Spot        Sandwich Place   
28                      Pizza Place               Brewery   
29          State / Provincial Park                  Café   
35                             Park           Pizza Place   
36                             Café               Brewery   
38              American Restaurant    Mexican Restaurant   
41               Seafood Restaurant   American Restaurant   
43               Mexican Restaurant           Pizza Place   
44                      Pizza Place        Ice Cream Shop   
47                  

##### There are four other locations that are similar to Flagstaff<br>
1. Asheville, NC
2. Madison, WI
3. Athens, OH
4. Fort Collins, CO

Madison, WI does not have coffee shops as the top 10 most common Venues. <br>Now, we need to look for the best location to open a coffee shop in Madison, WI.

# Explore Asheville, NC

### Find all zip codes in Ashville

In [34]:
search = SearchEngine(simple_zipcode=True)

In [35]:
res = search.by_city_and_state('Asheville', 'NC')
[x.zipcode for x in res]

['28801', '28803', '28804', '28805', '28806']

In [36]:
ash_zips = [x.zipcode for x in res]

### Find all places that serve coffee

In [37]:
def getVenuesByZip(zipcodes):
    count = 1
    venues_list=[]
    for zipcode in zipcodes:
        print(zipcode, count)
        count += 1
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&near={}&limit={}&query={}'.format(
                    CLIENT_ID, 
                    CLIENT_SECRET, 
                    VERSION, 
                    zipcode,
                    LIMIT,
                    'coffee')

        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(x['venue']['name'],
                            x['venue']['location']['lat'], 
                            x['venue']['location']['lng'],
                            zipcode,
                            x['venue']['categories'][0]['name']) for x in results])

        nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
        nearby_venues.columns = ['Venue',
                                'Venue Latitude', 
                                'Venue Longitude',
                                'Zip Code',
                                'Venue Category']
        
    return(nearby_venues)

In [38]:
ash_venues = getVenuesByZip(ash_zips)

28801 1
28803 2
28804 3
28805 4
28806 5


In [39]:
ash_venues

Venue  Venue Latitude  Venue Longitude  \
0             French Broad Chocolate Lounge       35.595231       -82.551178   
1                               Posana Cafe       35.594699       -82.551753   
2                            Corner Kitchen       35.566289       -82.541270   
3                          High Five Coffee       35.595801       -82.553448   
4                      High Five Coffee Bar       35.601041       -82.555424   
5                                 Dobra Tea       35.597427       -82.553548   
6              Double D's Coffee & Desserts       35.593223       -82.551365   
7                                      Hole       35.580767       -82.573272   
8                          Vortex Doughnuts       35.589190       -82.553733   
9                         Battle Cat Coffee       35.577951       -82.578697   
10                    Trade and Lore Coffee       35.594782       -82.555846   
11                                Starbucks       35.601549       -82.547024   
12                            Twisted Crepe       35.596422       -82.555245   
13                    Old Europe Pâtisserie       35.595775       -82.552032   
14                          Ultra Coffeebar       35.585018       -82.564694   
15                                Starbucks       35.566438       -82.544700   
16                      PennyCup Coffee Co.       35.581156       -82.564001   
17                              Well Played       35.594218       -82.556246   
18        The Green Sage Coffeehouse & Cafe       35.595576       -82.552028   
19                                  The Rhu       35.594549       -82.552604   
20                                    Bomba       35.595010       -82.551931   
21                           Dunkin' Donuts       35.613031       -82.554958   
22                 Green Sage Cafe Westgate       35.592598       -82.575205   
23                             World Coffee       35.595137       -82.555233   
24                         True Confections       35.595641       -82.556142   
25                        Izzy's Coffee Den       35.597329       -82.553493   
26                            Tod's Tasties       35.598988       -82.561816   
27                            Dunkin Donuts       35.569719       -82.544890   
28                              Duck Donuts       35.604978       -82.553566   
29                      Pennycup Coffee Co.       35.593756       -82.550157   
30                                  Cafe 64       35.596435       -82.555259   
31                               Noble Kava       35.593750       -82.550859   
32                       Wall Street Coffee       35.594478       -82.556337   
33                                  the Hub       35.577413       -82.575124   
34                            The Waterbird       35.606487       -82.546035   
35              Liberty House Coffee & Café       35.602445       -82.551492   
36                     The Asheville Public       35.588040       -82.564577   
37                            Summit Coffee       35.577488       -82.566890   
38           World Coffee Cafe' and Sky Bar       35.595190       -82.555150   
39                         10th Muse Coffee       35.565384       -82.538447   
40                                   Izzy's       35.577929       -82.578723   
41                   Inn On Biltmore Estate       35.557834       -82.580520   
42                         Sunny Point Cafe       35.578622       -82.588790   
43                           Corner Kitchen       35.566289       -82.541270   
44                                     Hole       35.580767       -82.573272   
45                        Battle Cat Coffee       35.577951       -82.578697   
46                                Starbucks       35.483142       -82.553702   
47                                Starbucks       35.523724       -82.529097   
48                                Starbucks       35.566438       -82.544700   
49                           The Green

### Map out all spots that serve coffee

In [40]:
# Get Latitude and logitude for Asheville, NC
API_KEY = 'AIzaSyDfaz-Hluid0RfZy6NDa3z9l26Hcuws93M'
gmaps = googlemaps.Client(key=API_KEY)
geocode_result = gmaps.geocode('Asheville, NC')
latitude = geocode_result[0]['geometry']['location']['lat']
longitude = geocode_result[0]['geometry']['location']['lng']

In [41]:
np.arange(len(ash_venues['Zip Code'].unique()))

array([0, 1, 2, 3, 4])

In [42]:
# create map
map_venues = folium.Map(location=[latitude, longitude], zoom_start=12)
# set color scheme for the clusters
zip_dic = {i:x+1 for x,i in enumerate(ash_venues['Zip Code'].unique())}
x = np.arange(len(zip_dic))
ys = [i+x+(i*x)**2 for i in range(len(zip_dic))]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]


# add markers to the map
markers_colors = []
for lat, lon, venue, zipcode in zip(ash_venues['Venue Latitude'], ash_venues['Venue Longitude'], ash_venues['Venue'], ash_venues['Zip Code']):
    label = folium.Popup(str(venue) + ' Zip Code: ' + str(zipcode), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[zip_dic[zipcode]-1],
        fill=True,
        fill_color=rainbow[zip_dic[zipcode]-1],
        fill_opacity=0.7).add_to(map_venues)

map_venues

Observations:
1. University of North Carolina, Asheville is located in zip code 28804
2. There is a lot of competition in the 28801 area coe
3. 28805 and 28803 seem to be far from the University.
4. There aren't many coffee shops in area code 28804. Seems like the best place to open that is near the university